torch/csrc/distributed/c10d/ProcessGroupUCC.cpp

```diff
@@ -0,0 +1,1870 @@
+  // main thread will throw out the exception then. There is no "abort"
+  // function in UCC currently.
+  while (!isCompleted())
+    ;
+  setAndThrowException();
+  // manually call profiling end callbacks if they are set,
+  // since progress thread does not own WorkUCC
+  if (ProcessGroup::Work::recordFunctionEndCallback_) {
+    ProcessGroup::Work::recordFunctionEndCallback_();
+    ProcessGroup::Work::recordFunctionEndCallback_ = nullptr;
+  }
+  return true;
+}
+
+c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupUCC::WorkUCC::getFuture() {
+  return future_;
+}
+
+std::vector<at::Tensor> ProcessGroupUCC::WorkUCC::result() {
+  return *outputs_;
+}
+
+void ProcessGroupUCC::ProgressEntry::finalize(std::exception_ptr eptr) {
+  ucc_status_t status = UCC_OK;
+
+  if (request_ != nullptr) {
+    status = request_->status;
+    comm_->free_request(request_);
+  }
+  if (eptr) {
+    eptr_ = eptr;
+  } else {
+    status_ = status;
+  }
+  if (future_) {
+    if (eptr) {
+      future_->setError(eptr);
+    } else {
+      future_->markCompleted(
+          c10::IValue(data ? data->dst : std::vector<at::Tensor>()));
+    }
+  }
+}
+
+Comm::Comm(
+    const c10::intrusive_ptr<ProcessGroupUCCLogger>& logger_,
+    std::shared_ptr<torch_ucc_oob_coll_info_t> oob_,
+    c10::Device dev,
+    bool is_health_check)
+    : logger(logger_),
+      oob(oob_),
+      ucx_comm(oob->size, logger),
+      ucc_comm(oob, logger),
+      finalize_phase(
+          is_health_check ? TORCH_UCC_HEALTH_CHECK : TORCH_UCC_FINALIZE),
+      cuda_device_index(TORCH_UCC_DEVICE_NOT_SET) {
+  if (dev.is_cuda()) {
+    cuda_device_index = dev.index();
+  }
+  stop_progress_loop = false;
+  collective_inprogress = false;
+  progress_thread = std::thread(&Comm::progress_loop, this);
+#ifdef _GNU_SOURCE
+  pthread_setname_np(progress_thread.native_handle(), "ucc-progress");
+#endif
+}
+
+Comm::~Comm() {
+  std::unique_lock<std::mutex> lock(mutex);
+  queue_consume_cv.wait(
+      lock, [&] { return progress_queue.empty() && !collective_inprogress; });
+  stop_progress_loop = true;
+  lock.unlock();
+  queue_produce_cv.notify_all();
+  progress_thread.join();
+}
+
+std::shared_ptr<Comm> Comm::get_comm(
+    uint32_t& id,
+    c10::Device dev,
+    std::shared_ptr<torch_ucc_oob_coll_info_t> oob,
+    const c10::intrusive_ptr<ProcessGroupUCCLogger>& logger,
+    bool is_health_check) {
+  static std::mutex m;
+  static std::weak_ptr<Comm> comm;
+  static uint32_t comm_id;
+
+  std::lock_guard<std::mutex> lock(m);
+  id = (comm_id % TORCH_UCX_MAX_COMM);
+
+  std::string group_id = "group_id";
+  if (is_health_check) {
+    group_id = c10::str(dev.type()) + "/" + group_id;
+  }
+
+  std::vector<uint8_t> remote_comm_id;
+  oob->store->deleteKey(group_id + std::to_string(0));
+  if (oob->rank != 0) {
+    std::vector<uint8_t> val = std::vector<uint8_t>(
+        reinterpret_cast<uint8_t*>(&id),
+        reinterpret_cast<uint8_t*>(&id) + sizeof(id));
```

Comment:
nit: This is repeated quite a lot throughout this file and we should probably add a function overload for store that just takes a string and performs this conversion internally. This is more on the PT-D team to do.